### PR TITLE
WebGPURenderer: Fix `vec2` and `vec3` for `storageObject` in StorageBufferNode

### PR DESCRIPTION
**Pull request by RenaudRohlinger**
I've made some significant updates to how we handle data packing into textures for Pixel Buffer Object fallback of storage buffers.
Previously, we were tackling this with RGBA channels, but it turned out to be broken and doing like so would make things more complex than necessary, especially with overlapping values. For example when extracting vec2 and vec3 from texels, with vec3 data overlapping across texels making the code hard to maintain and scale.

To make things easier we're now packing data directly into their proper formats (Red, RG, RGB, RGBA), which makes the code much cleaner and more straightforward for all cases. Might even bring a small boost performance.

Also updated the example to improve E2E testing:
[https://raw.githack.com/renaudrohlinger/three.js/improve-test/examples/webgpu_storage_buffer.html](https://raw.githack.com/renaudrohlinger/three.js/improve-test/examples/webgpu_storage_buffer.html
)Screenshot 2024-02-08 at 0 15 34

_This contribution is funded by [Utsubo](https://utsubo.com/)_

### CODE REVIEW

**Comment by LeviPesin** [comment]
Hm... Would it be beneficial to re-add the example from https://github.com/mrdoob/three.js/pull/25273?

**Comment by sunag** [comment]
What do you think about guide the user to create buffers with padding similar to what happens in std140 layout, and deal with `float` and `vec4` in the buffer for example, perhaps it would simplify the process?

**Comment by RenaudRohlinger** [comment]
Hum, I think it would complexify things for developers as difficult as using interleaved array buffers (no one like these things 😄).
I don't see the downside of using the native formats made available by graphics cards but I do see the downsides of having to guide users for a specific way of packing up their datas in order to support only a specific WebGL fallback.

Also maybe I'm missing something here?

By the way is it antipattern to try to read a buffer through `arraybuffernode.element(instanceIndex)` in a `vertex` when it was previously used as storage buffer in a `compute` shader?
I wanted to implement a test here by using `arrayBufferNodes[ 0 ].element( instanceIndex )` in a vertex shader but got this error:
```
Read-write storage buffer binding is used with a visibility (ShaderStage::Vertex) that contains ShaderStage::Vertex (note that read-only storage buffer bindings are allowed).
 - While validating entries[0]
 - While validating [BindGroupLayoutDescriptor]
 - While calling [Device].CreateBindGroupLayout([BindGroupLayoutDescriptor]).
 ```
 Or should we add something like `arrayBufferNodes[ 0 ].element( instanceIndex ).visibility('read')` or `.read()` or just fix something in the WebGPU Backend?
 
 Currently all examples are using `toAttribute()` but `WebGL` will probably not like it. I could try to investigate on that support (some rebinding post-PBO of the array buffer when calling `toAttribute()` I guess?)

**Comment by sunag** [comment]
I imagined something simpler for this, similar to the way `RangeNode` deals with buffers for example. Maybe the way I explained it is to understand something much more difficult than it really is, using `vec4` instead of `vec3` to get more performance I think is a benefit.

**Comment by sunag** [comment]
@RenaudRohlinger Is this happening to you too? It seems like a loss of data on the part of the WebGPU.

image

**Comment by sunag** [comment]
> Currently all examples are using toAttribute() but WebGL will probably not like it. I could try to investigate on that support (some rebinding post-PBO of the array buffer when calling toAttribute() I guess?)

I think an intermediate Node in `toAttribute` could solve this by handling `setup()`. Maybe you already have a minimal example for share with us? I could be helping with this implementation....

File: examples/jsm/renderers/webgl/nodes/GLSLNodeBuilder.js [hunk 1]
```diff
@@ -94,20 +94,36 @@ ${ flowData.code }
...
+			let format = RedFormat;
 
-			const newSize = width * height * 4; // 4 floats per pixel due to RGBA format
+			if ( itemSize === 2 ) {
+
+				format = RGFormat;
+
+			} else if ( itemSize === 3 ) {
+
+				format = 6407; // patch since legacy doesn't use RGBFormat for rendering but here it's needed for packing optimization
```
Comment (LeviPesin) [inline examples/jsm/renderers/webgl/nodes/GLSLNodeBuilder.js:106]: @Mugen87 Maybe we can introduce the `RGBFormat` constant back?

**Comment by RenaudRohlinger** [comment]
Thanks @sunag! I actually just set back the `dualAttributeData.switchBuffers();` to always happen even in `PBO` setup which fixes all the issues `toAttribute()` included by re-binding the correct array to the program.

> Maybe you already have a minimal example for share with us?
Try to replace any compute example (`webgpu_compute_particles_rain` for example) from `storage()` to `storageObject()` on dev and they will all break.
This PR fixe all these issues, now you can safely use `PBO` via `storageObject` on any example and they will work just fine.

Regarding the decision to replace vec3 with vec4, I don't quite understand the logic behind it in terms of optimization. Using vec4 will only result in a larger ArrayBuffer and consume more memory unnecessarily if the user needs to utilize a vec3.
I would not advise to drop `vec3` support to `storage_buffer` especially for `WebGL that tries to fetch other elements` fallback support only (very rare case).
The only justification I could see for not using RGB is that GPUs might fill the last texel, thereby treating it as RGBA, which isn't a significant issue considering the inefficiency of trying to pack and unpack vec3.

About the precision issue in the WebGPU side I had it when using non power-of-2 array buffers. Maybe the issue is GPU related and related to how I weirdly fetch indexes in `colorNode` (`const index = uint( uv().x.mul( size ).floor() ).toVar();`).

**Review by sunag — APPROVED** [review]

**Pull request merged**